com.unity.netcode.gameobjects/Runtime/Messaging/Messages/RpcMessages.cs

```diff
@@ -0,0 +1,146 @@
+            }
+
+            writer.WriteValue(metadata);
+            writer.WriteBytes(payload.GetUnsafePtr(), payload.Length);
+        }
+
+        public static unsafe bool Deserialize(ref FastBufferReader reader, ref NetworkContext context, ref RpcMetadata metadata, ref FastBufferReader payload)
+        {
+            int metadataSize = FastBufferWriter.GetWriteSize<RpcMetadata>();
+            if (!reader.TryBeginRead(metadataSize))
+            {
+                throw new InvalidOperationException("Not enough data in the buffer to read RPC meta.");
+            }
+
+            reader.ReadValue(out metadata);
+
+            var networkManager = (NetworkManager)context.SystemOwner;
+            if (!networkManager.SpawnManager.SpawnedObjects.ContainsKey(metadata.NetworkObjectId))
+            {
+                networkManager.SpawnManager.TriggerOnSpawn(metadata.NetworkObjectId, reader, ref context);
+                return false;
+            }
+
+            var networkObject = networkManager.SpawnManager.SpawnedObjects[metadata.NetworkObjectId];
+            var networkBehaviour = networkManager.SpawnManager.SpawnedObjects[metadata.NetworkObjectId].GetNetworkBehaviourAtOrderIndex(metadata.NetworkBehaviourId);
+            if (networkBehaviour == null)
+            {
+                return false;
+            }
+
+            if (!NetworkManager.__rpc_func_table.ContainsKey(metadata.NetworkRpcMethodId))
+            {
+                return false;
+            }
+
+            payload = new FastBufferReader(reader.GetUnsafePtr(), Allocator.None, reader.Length - metadataSize, metadataSize);
+
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            if (NetworkManager.__rpc_name_table.TryGetValue(metadata.NetworkRpcMethodId, out var rpcMethodName))
+            {
+                networkManager.NetworkMetrics.TrackRpcReceived(
+                    context.SenderId,
+                    networkObject,
+                    rpcMethodName,
+                    networkBehaviour.__getTypeName(),
+                    reader.Length);
+            }
+#endif
+
+            return true;
+        }
+
+        public static void Handle(ref NetworkContext context, ref RpcMetadata metadata, ref FastBufferReader payload, ref __RpcParams rpcParams)
+        {
+            var networkManager = (NetworkManager)context.SystemOwner;
+            var networkObject = networkManager.SpawnManager.SpawnedObjects[metadata.NetworkObjectId];
+            var networkBehaviour = networkObject.GetNetworkBehaviourAtOrderIndex(metadata.NetworkBehaviourId);
+
+            NetworkManager.__rpc_func_table[metadata.NetworkRpcMethodId](networkBehaviour, payload, rpcParams);
+        }
+    }
+
+    internal struct RpcMetadata
+    {
+        public ulong NetworkObjectId;
+        public ushort NetworkBehaviourId;
+        public uint NetworkRpcMethodId;
+    }
+
+    internal struct ServerRpcMessage : INetworkMessage
```

Comment:
In addition to going splitting into _begin_, _send_ and _end_, this splits RpcMessage into ServerRpcMessage and ClientRpcMessage. I'm just curious if you could elaborate on the rationale behind this change ? Are the two different enough to warrant separate classes, why ? What's the advantage ? (Not arguing against, just trying get an idea or the reasoning)

Reply:
quoting from the PR description here:
> we no longer need to serialize rpc type (server/client) since the message type itself will be used as the identifier.

`INetworkMessage` types are being registered automatically for us which means there's already a type-based dispatch going on within `MessagingSystem`. We don't need to replicate the same behavior here with an `enum RpcType { Server, Client }` here.